structuretoolkit/analyse/snap.py

```diff
@@ -0,0 +1,397 @@
+        if v in bispec_options
+    }
+    kw_substrings = [f"{k} {v}" for k, v in kw_options.items()]
+    kwargs = " ".join(kw_substrings)
+
+    for op, base in zip(("b", "db", "vb"), (base_b, base_db, base_vb)):
+        # print("Setting up compute",op)
+        command = f"{base} {radelem} {wj} {kwargs}"
+        # print(command)
+        lmp.command(command)
+
+    for cname in ["b", "db", "vb"]:
+        lmp.command(f"compute {cname}_sum all reduce sum c_{cname}[*]")
+
+
+def extract_computes_snappy(lmp, num_atoms, n_coeff, num_types):
+    lmp_atom_ids = lmp.numpy.extract_atom_iarray("id", num_atoms).flatten()
+    assert np.all(
+        lmp_atom_ids == 1 + np.arange(num_atoms)
+    ), "LAMMPS seems to have lost atoms"
+
+    # Extract types
+    lmp_types = lmp.numpy.extract_atom_iarray(name="type", nelem=num_atoms).flatten()
+    lmp_volume = lmp.get_thermo("vol")
+
+    # Extract Bsum
+    lmp_bsum = extract_compute_np(lmp, "b_sum", 0, 1, (n_coeff))
+
+    # Extract B
+    lmp_barr = extract_compute_np(lmp, "b", 1, 2, (num_atoms, n_coeff))
+
+    type_onehot = np.eye(num_types)[lmp_types - 1]  # lammps types are 1-indexed.
+    # has shape n_atoms, n_types, num_coeffs.
+    # Constructing so it has the similar form to db and vb arrays. This adds some memory usage,
+    # but not nearly as much as vb or db (which are factor of 6 and n_atoms*3 larger, respectively)
+
+    b_atom = type_onehot[:, :, np.newaxis] * lmp_barr[:, np.newaxis, :]
+    b_sum = b_atom.sum(axis=0) / num_atoms
+
+    try:
+        assert np.allclose(
+            lmp_bsum, lmp_barr.sum(axis=0), rtol=1e-12, atol=1e-12
+        ), "b_sum doesn't match sum of b"
+    except AssertionError:
+        print(lmp_bsum)
+        print(lmp_barr.sum(axis=0))
+
+    lmp_dbarr = extract_compute_np(lmp, "db", 1, 2, (num_atoms, num_types, 3, n_coeff))
+    lmp_dbsum = extract_compute_np(lmp, "db_sum", 0, 1, (num_types, 3, n_coeff))
+    assert np.allclose(
+        lmp_dbsum, lmp_dbarr.sum(axis=0), rtol=1e-12, atol=1e-12
+    ), "db_sum doesn't match sum of db"
+    db_atom = np.transpose(lmp_dbarr, (0, 2, 1, 3))
+
+    lmp_vbarr = extract_compute_np(lmp, "vb", 1, 2, (num_atoms, num_types, 6, n_coeff))
+    lmp_vbsum = extract_compute_np(lmp, "vb_sum", 0, 1, (num_types, 6, n_coeff))
+    assert np.allclose(
+        lmp_vbsum, lmp_vbarr.sum(axis=0), rtol=1e-12, atol=1e-12
+    ), "vb_sum doesn't match sum of vb"
+    vb_sum = np.transpose(lmp_vbsum, (1, 0, 2)) / lmp_volume * eV_div_A3_to_bar
+
+    dbatom_shape = db_atom.shape
+    vbsum_shape = vb_sum.shape
+    a_fit = np.concatenate(
+        (
+            b_sum,
+            db_atom.reshape(
+                dbatom_shape[0] * dbatom_shape[1], dbatom_shape[2] * dbatom_shape[3]
+            ),
+            vb_sum.reshape(vbsum_shape[0] * vbsum_shape[1], vbsum_shape[2]),
+        ),
+        axis=0,
+    )
+    return np.concatenate(
+        (np.array([np.eye(a_fit.shape[0])[0]]).T, a_fit), axis=1
+    ).copy()
```

Comment:
```suggestion
    )
```
Now I'm not sure anymore, but I'd have guessed that concatenate always gives a copy anyway.

Reply:
In general this is correct but here the entries of the numpy array are links to the values stored in the LAMMPS executable - to prevent duplicating the memory - so we have to copy them.